Adding classes to calculate the space charge distortions and corrections

matthias-kleiner opened the pull request:
The distortions and corrections can be calculated by using the potential
from a 3D histogram or an analytical formula as an input. These classes are
much faster and consistent than the current methods located in GPU/TPCSpaceChargeBase and will replace them.

matthias-kleiner committed:
Adding classes to calculate the space charge distortions and corrections

The distortions and corrections can be calculated by using the potential
from a 3D histogram or an analytical formula as an input. These classes are
much faster and consistent than the old methods located in GPU/TPCSpaceChargeBase.
The old folder (GPU/TPCSpaceChargeBase) where all the previous classes were stored are deleted
and all other classes and macros which used the old spacecharge classes
are modified to be able to use the new spacecharge classes.

wiechula commented on Detectors/TPC/spacecharge/include/TPCSpacecharge/PoissonSolver.h:
```diff
@@ -0,0 +1,487 @@
+  /// Helper function to check if the integer is equal to a power of two
+  /// \param i the number
+  /// \return 1 if it is a power of two, else 0
+  bool isPowerOfTwo(int i) const;
```
This could be simply
```
return ((i>0) && !(i & (i - 1)));
```

wiechula commented on Detectors/TPC/spacecharge/include/TPCSpacecharge/PoissonSolverStructs.h:
```diff
@@ -0,0 +1,170 @@
+  /// operator to set the values
+  DataT& operator()(const unsigned int iR, const unsigned int iZ, const unsigned int iPhi)
+  {
+    return mStorage[iR + mNr * (iZ + mNz * iPhi)];
```
why not use  `getIndex`

wiechula commented on Detectors/TPC/spacecharge/include/TPCSpacecharge/PoissonSolverStructs.h:
```diff
@@ -0,0 +1,170 @@
+  /// operator to read the values
+  const DataT& operator()(const unsigned int iR, const unsigned int iZ, const unsigned int iPhi) const
+  {
+    return mStorage[iR + mNr * (iZ + mNz * iPhi)];
```
why not use  `getIndex`

wiechula commented on Detectors/TPC/spacecharge/include/TPCSpacecharge/PoissonSolverStructs.h:
```diff
@@ -0,0 +1,170 @@
+/// this is a simple vector class which is used in the poisson solver class
+
+/// \tparam DataT the data type of the mStorage which is used during the calculations
+template <typename DataT = double>
```
Why is this not in an own header as e.g. DataContainer3D?

wiechula commented on Detectors/TPC/spacecharge/include/TPCSpacecharge/SpaceChargeStructs.h:
```diff
@@ -0,0 +1,286 @@
+#include <cmath>
+#include "TPCSpacecharge/TriCubic.h"
+#include "DataFormatsTPC/Defs.h"
+#include "TPCSpacecharge/SpaceChargeStructs.h"
```
remove?

wiechula commented on Detectors/TPC/spacecharge/include/TPCSpacecharge/SpaceChargeStructs.h:
```diff
@@ -0,0 +1,286 @@
+
+  static constexpr unsigned int getID() { return ID; }
+
+  // private:
```
why commented?

wiechula commented on Detectors/TPC/spacecharge/include/TPCSpacecharge/Vector.h:
```diff
@@ -0,0 +1,173 @@
+
+ public:
+  /// default constructor
+  Vector() {}
```
`=default`

wiechula commented on Detectors/TPC/spacecharge/src/SpaceCharge.cxx:
```diff
@@ -0,0 +1,1007 @@
+void SpaceCharge<DataT, Nz, Nr, Nphi>::calcGlobalDistWithGlobalCorrIterative(const DistCorrInterpolator<DataT, Nz, Nr, Nphi>& globCorr, const int maxIter, const DataT approachZ, const DataT approachR, const DataT approachPhi, const DataT diffCorr)
+{
+  // for nearest neighbour search see: https://doc.cgal.org/latest/Spatial_searching/Spatial_searching_2searching_with_point_with_info_8cpp-example.html
+  typedef CGAL::Exact_predicates_inexact_constructions_kernel Kernel;
+  typedef Kernel::Point_3 Point_3;
+  typedef boost::tuple<Point_3, std::array<unsigned int, 3>> Point_Index;
+  typedef CGAL::Search_traits_3<Kernel> Traits_base;
+  typedef CGAL::Search_traits_adapter<Point_Index, CGAL::Nth_of_tuple_property_map<0, Point_Index>, Traits_base> Traits;
+  typedef CGAL::Orthogonal_k_neighbor_search<Traits> K_neighbor_search;
+  typedef K_neighbor_search::Tree Tree;
```
change to `using`?

wiechula reviewed (changes requested):
Perhaps instead of `...Structs.h` you could use `...Helpers.h`. In some of the file you declare structs with private members. I think I would rather make them classes.

Otherwise only minor comments

ehellbar commented on Detectors/TPC/spacecharge/include/TPCSpacecharge/PoissonSolverStructs.h:
```diff
@@ -0,0 +1,170 @@
+
+template <typename DataT = double>
+struct TPCParameters {
+  static constexpr DataT TPCZ0{249.525};                        ///< nominal gating grid position
```
should be nominal G1T position in the comment

ehellbar commented on Detectors/TPC/spacecharge/src/PoissonSolver.cxx:
```diff
@@ -0,0 +1,1505 @@
+void PoissonSolver<DataT, Nz, Nr, Nphi>::poissonSolver2D(DataContainer& matricesV, const DataContainer& matricesCharge)
+{
+  poissonMultiGrid2D(matricesV, matricesCharge);
+}
```
Just a note:
iPhi (third argument of the function) is set to 0 by default. In the function itself, only the iPhi bin of the output matrix matricesV will be filled (line 181). So for the 2D solution, only 1 phi bin is calculated and filled. Once (if) we start to use the 2D solutions for the ion drift, we have to revisit this.

ehellbar commented on Detectors/TPC/spacecharge/test/testO2TPCPoissonSolver.cxx:
```diff
@@ -0,0 +1,240 @@
+  const o2::tpc::AnalyticalFields<DataT> analyticalFields;
+  // set the boudnary and charge for numerical poisson solver
+  setChargeDensityFromFormula<DataT, Nz, Nr, Nphi>(analyticalFields, grid3D, charge);
+  setPotentialBoundaryFromFormula<DataT, Nz, Nr, Nphi>(analyticalFields, grid3D, potentialNumerical);
```
Is there a reason why you set the boundary potentials of the numerical calculation with the values from the analytical formula? The numerical solution should work fine without this as it should calculate the potential at these boundaries from the charge? 

The function to set only the boundary potentials is supposed to provide to possibility to account for irregularities of the nominal potentials, e.g. charge-up effects at dielectric sector boundaries.

matthias-kleiner replied:
> Is there a reason why you set the boundary potentials of the numerical calculation with the values from the analytical formula?

It was like this in the original code. So I always used theses as the boundary conditions.

> The numerical solution should work fine without this as it should calculate the potential at these boundaries from the charge?

I am not sure if boundary conditions are really needed or not for this method. I made a quick comparison by setting the boundary potential and then calculating the potential with the poisonsolver and without setting the boundary potential. The difference is quit big between the resulting potential to the reference potential. See the two plots "analytical potential - numerical potential". The first plot is with the boundary conditions and the setting plot without these boundary conditions for the potential.


![with_boundary_conditions](https://user-images.githubusercontent.com/48915672/97005999-7b425600-153f-11eb-9b76-9ca79ac615af.png)
![no_boundary_conditions](https://user-images.githubusercontent.com/48915672/97006055-8e552600-153f-11eb-80bb-618898edcb98.png)

ehellbar replied:
Ok. Now I am thinking about what this means for the case when we use density histograms as input. Maybe at some point we can try to put there the nominal TPC boundary conditions (potential CE, G1T and field cage) already for the Poisson solver instead only adding it (Ez0=400 V/cm) for the calculation of the distortions.

ehellbar commented:
Hello @matthias-kleiner ,

can you add a method to the main class which calculates all the quantities (potential, E-field, distortion, correction), similar to the function calculateDistortionsCorrections of the example macro? It would make the application of the class in future studies much simpler, or what do you think?

Otherwise, it looks very good to me.

Cheers,
Ernst

matthias-kleiner commented:
> Hello @matthias-kleiner ,
> 
> can you add a method to the main class which calculates all the quantities (potential, E-field, distortion, correction), similar to the function calculateDistortionsCorrections of the example macro? It would make the application of the class in future studies much simpler, or what do you think?
> 
> Otherwise, it looks very good to me.
> 
> Cheers,
> Ernst

Hi Ernst, 
I just implemented such a method and use this method now in the macro.

I also implemented all the other comments from you and Jens.

matthias-kleiner commented:
@ehellbar 

I still have to invert the distortions in z direction for the A-(?)Side. Should I do it directly during the filling of the values or when quering the values (like it was done in the old code). I assume it should be done during the filling?

And there is in the TPCSimulation folder a class with the same name "SpaceCharge.h", which causes some problems when running a simulation.

ehellbar commented:
Hi @matthias-kleiner ,

> I still have to invert the distortions in z direction for the A-(?)Side. Should I do it directly during the filling of the values or when quering the values (like it was done in the old code). I assume it should be done during the filling?

I think it concerns the values on the C side? On the A side, the direction along z in the array/matrix entries is 0 -> nZ and for the z coordinate 0 -> 250 cm.  On the C side it is 0 -> nZ and 0 -> -250 cm. Differentiation / integration along z is done using the array / matrix index, right? So on the C side, you will basically invert the z axis (direction -z instead of z) which results in a change of the sign. I think this happens first when calculating the Ez field from the potential? So you can change the sign of Ez on the C side and it should be correct for all following steps? But we should check the signs of the E fields, local and global distortions after the change.

> And there is in the TPCSimulation folder a class with the same name "SpaceCharge.h", which causes some problems when running a simulation.

Is it related to the new classes? They are not yet used in the simulation and are in a different scope, right (e.g. TPCSpaceCharge instead of TPCSimulation)?

Cheers,
Ernst

matthias-kleiner commented:
Hi @ehellbar 
> Is it related to the new classes? They are not yet used in the simulation and are in a different scope, right (e.g. TPCSpaceCharge instead of TPCSimulation)?

when running a simulation it completes, but I get some error from I think ROOT:
`error: redefinition of 'SpaceCharge' as different kind of symbol`
`note: previous definition is here`
`namespace o2{namespace tpc{class __attribute__((annotate("$clingAutoload$TPCSimulation/SpaceCharge.h")))  SpaceCharge;}}`
I think its because they are in the same namespace `o2::tpc`

matthias-kleiner commented:
> Hi @matthias-kleiner ,
> 
> > I still have to invert the distortions in z direction for the A-(?)Side. Should I do it directly during the filling of the values or when quering the values (like it was done in the old code). I assume it should be done during the filling?
> 
> I think it concerns the values on the C side?

Hi @ehellbar,
you are right. Its the value on the C-Side.

> So you can change the sign of Ez on the C side and it should be correct for all following steps? But we should check the signs of the E fields, local and global distortions after the change.

I just checked it, but the resulting corrections/distortions are slightly different after changing the sign of the ez field. I think changing the sign of Ez also changes these integrals "int er/ez dz" etc. when calculating the distortions. So i think one should integrate then from 0 -> -nZBins, but then I have to change several other things too.

ehellbar commented:
Hello @matthias-kleiner ,
yes, you are right. You have to check the sign everywhere where the Ez is used. In the integrals Er/(ezField+Ez), Ephi/(ezField+Ez) one has to change also the sign of ezField accordingly, as it has negative sign on the A side and positive sign on the C side (?).

ehellbar commented:
> when running a simulation it completes, but I get some error from I think ROOT:
> error: redefinition of 'SpaceCharge' as different kind of symbol
> note: previous definition is here
> namespace o2{namespace tpc{class __attribute__((annotate("$clingAutoload$TPCSimulation/SpaceCharge.h"))) SpaceCharge;}}
> I think its because they are in the same namespace o2::tpc
> 

The class is mainly used in the TPC digitizer and some macros.
```
(base) hellbaer@hellbaer-Thinkpad:~/software/alice/o2-dev/O2$ grep -nr "TPCSimulation/SpaceCharge" *
Detectors/TPC/simulation/src/SpaceCharge.cxx:28:#include "TPCSimulation/SpaceCharge.h"
Detectors/TPC/simulation/CMakeLists.txt:42:                                  include/TPCSimulation/SpaceCharge.h)
Detectors/TPC/simulation/include/TPCSimulation/Digitizer.h:21:#include "TPCSimulation/SpaceCharge.h"
Detectors/TPC/simulation/macro/createResidualDistortionObject.C:25:#include "TPCSimulation/SpaceCharge.h"
GPU/TPCFastTransformation/macro/generateTPCCorrectionNTuple.C:27:#include "TPCSimulation/SpaceCharge.h"
```
Maybe the best solution at the moment is to replace the usage of the other SpaceCharge class by your code, which requires a couple of new methods used in the other SpaceCharge class: distortElectron and correctElectron and an enum class SCDistortionType. Then, the old class can be removed from compilation. It still contains methods to propagate the space-charge density or ions which we should continue to develop and move to the spacecharge folder, either as a new class or as addition to SpaceCharge. But once the dependencies on the old SpaceCharge class are replaced by yours, we can basically delete the old one (getting the ion movement methods later from the history) and the GPU/TPCSpaceChargeBase folder as well.

matthias-kleiner commented:
> > when running a simulation it completes, but I get some error from I think ROOT:
> > error: redefinition of 'SpaceCharge' as different kind of symbol
> > note: previous definition is here
> > namespace o2{namespace tpc{class **attribute**((annotate("$clingAutoload$TPCSimulation/SpaceCharge.h"))) SpaceCharge;}}
> > I think its because they are in the same namespace o2::tpc
> 
> The class is mainly used in the TPC digitizer and some macros.
> 
> ```
> (base) hellbaer@hellbaer-Thinkpad:~/software/alice/o2-dev/O2$ grep -nr "TPCSimulation/SpaceCharge" *
> Detectors/TPC/simulation/src/SpaceCharge.cxx:28:#include "TPCSimulation/SpaceCharge.h"
> Detectors/TPC/simulation/CMakeLists.txt:42:                                  include/TPCSimulation/SpaceCharge.h)
> Detectors/TPC/simulation/include/TPCSimulation/Digitizer.h:21:#include "TPCSimulation/SpaceCharge.h"
> Detectors/TPC/simulation/macro/createResidualDistortionObject.C:25:#include "TPCSimulation/SpaceCharge.h"
> GPU/TPCFastTransformation/macro/generateTPCCorrectionNTuple.C:27:#include "TPCSimulation/SpaceCharge.h"
> ```
> 
> Maybe the best solution at the moment is to replace the usage of the other SpaceCharge class by your code, which requires a couple of new methods used in the other SpaceCharge class: distortElectron and correctElectron and an enum class SCDistortionType. Then, the old class can be removed from compilation. It still contains methods to propagate the space-charge density or ions which we should continue to develop and move to the spacecharge folder, either as a new class or as addition to SpaceCharge. But once the dependencies on the old SpaceCharge class are replaced by yours, we can basically delete the old one (getting the ion movement methods later from the history) and the GPU/TPCSpaceChargeBase folder as well.

Hi @ehellbar,
since the new SpaceCharge class is templated, I am not sure about the best way to pass the template parameters when one uses this class in the Digitizer. Of course I can set a default grid size, but then one looses the flexibility to change the gridsize during run time. I could just get rid of all the templates and use vectors everywhere, but that would probably require some modifications of the code. What do you think about it?

ehellbar commented:
Hi @matthias-kleiner ,
for the current use in the digitizer we can simply fix the grid size, e.g. to 129 x 129 x 180 (r x z x phi). The calculations with this grid size are quite fast (~10 sec?) and should be sufficiently precise. For studies e.g. of different grid sizes, we should use the class standalone. Fixing the grid size requires to also remove the option "gridSize" in the digitizer workflow O2/Steer/DigitizerWorkflow/src/TPCDigitizerSpec.cxx . Also, we have to check and modify the options initialSpaceChargeDensity (providing file name and histogram name) to calculate the distortion from a histogram and readSpaceCharge (provide a precalculated SpaceCharge object).

matthias-kleiner commented:
> Hi @matthias-kleiner ,
> for the current use in the digitizer we can simply fix the grid size, e.g. to 129 x 129 x 180 (r x z x phi). The calculations with this grid size are quite fast (~10 sec?) and should be sufficiently precise. For studies e.g. of different grid sizes, we should use the class standalone. Fixing the grid size requires to also remove the option "gridSize" in the digitizer workflow O2/Steer/DigitizerWorkflow/src/TPCDigitizerSpec.cxx . Also, we have to check and modify the options initialSpaceChargeDensity (providing file name and histogram name) to calculate the distortion from a histogram and readSpaceCharge (provide a precalculated SpaceCharge object).

Hi @ehellbar,

> for the current use in the digitizer we can simply fix the grid size, e.g. to 129 x 129 x 180 (r x z x phi)

Okay, I fixed it to 129x129x180.

> The calculations with this grid size are quite fast (~10 sec?)

When I run the external macro to calculate the distortions/corrections for an input histogram it takes ~10sec (8 Threads) for each side.

I replaced the old SpaceCharge class with the new one and it seems that it works in the digitizer. However when I simulate one event and then run the digitizer with
`o2-sim-digitizer-workflow -b --distortionType 2 --initialSpaceChargeDensity "/Users/matthias/alice/o2_macros/SpaceCharges/ATO-498/InputSCDensityHistograms_8000events.root,inputSCDensity3D_8000_avg"`
it picks up the histogram and calculates the distortions/corrections, but it looks like it performs the calculations multiple times if I dont run the digitizer with `--tpc-lanes 1`? (In the digitizer it takes around 28 seconds for both sides)

> readSpaceCharge (provide a precalculated SpaceCharge object)

So this means the SpaceCharge object is then read from a ROOT file or is it calculated somewhere else and then just passed to the function?

ehellbar commented:
> it picks up the histogram and calculates the distortions/corrections, but it looks like it performs the calculations multiple times if I dont run the digitizer with --tpc-lanes 1? (In the digitizer it takes around 28 seconds for both sides)

Yes, the way this is passed on at the moment, the distortions are calculated for each lane.

> So this means the SpaceCharge object is then read from a ROOT file or is it calculated somewhere else and then just passed to the function?

Yes, this is supposed to read a precalculated SpaceCharge object stored in a root file.

ehellbar commented:
Hello @matthias-kleiner ,
thanks for the updates.

Line 364 in TPCDigitizerSpec.cxx also should be removed:
`{"gridSize", VariantType::String, "129,144,129", {"Comma separated list of number of bins in (r,phi,z) for distortion lookup tables (r and z can only be 2**N + 1, N=1,2,3,...)"}},`

Can you also adjust the usage in GPU/TPCFastTransformation/macro/generateTPCCorrectionNTuple.C, please? This macro creates the fast correction objects using spline fits of the corrections, to be then applied in the reconstruction. 

Cheers,
Ernst

matthias-kleiner commented:
> > it picks up the histogram and calculates the distortions/corrections, but it looks like it performs the calculations multiple times if I dont run the digitizer with --tpc-lanes 1? (In the digitizer it takes around 28 seconds for both sides)
> 
> Yes, the way this is passed on at the moment, the distortions are calculated for each lane.
> 
> > So this means the SpaceCharge object is then read from a ROOT file or is it calculated somewhere else and then just passed to the function?
> 
> Yes, this is supposed to read a precalculated SpaceCharge object stored in a root file.

I just tried to implement a method to store the SpaceCharge file in a ROOT file, but for some reason it doesnt store the members. I think it might be related to all the static members which are used. But I already implemented a function to set the values (the DataContainer3D member) for the distortions/corrections etc. from a ROOT file. I think it should be easier to store the values for the distortions/corrections in a ROOT file and then set the data members from the ROOT file. Do you think this would also be fine? And if so which values have to be loaded then (only the global distortions and the global corrections)?

ehellbar commented:
Hi @matthias-kleiner ,
I also had some troubles with the streamers for the old class, and also I just streamed the lookup tables for the global corrections and distortions. So your solution sounds good to me.

matthias-kleiner commented:
> Hello @matthias-kleiner ,
> thanks for the updates.
> 
> Line 364 in TPCDigitizerSpec.cxx also should be removed:
> `{"gridSize", VariantType::String, "129,144,129", {"Comma separated list of number of bins in (r,phi,z) for distortion lookup tables (r and z can only be 2**N + 1, N=1,2,3,...)"}},`
> 
> Can you also adjust the usage in GPU/TPCFastTransformation/macro/generateTPCCorrectionNTuple.C, please? This macro creates the fast correction objects using spline fits of the corrections, to be then applied in the reconstruction.
> 
> Cheers,
> Ernst

hi @ehellbar,
So if  I understand correctly the distortions dont need to be calculated in this macro? So just the global corrections?
And is there some way to test the splines and check if everything is done correctly?
And is there some specific reason the SpaceCharge object is defined as a pointer above the function `(o2::tpc::SpaceCharge* sc = 0; )` instead as a simple object in the function `o2::tpc::SpaceCharge sc(mNRRows, mNPhiSlices, mNZCols);`?

ehellbar commented:
Hi @matthias-kleiner ,
in the macro generateTPCCorrectionNTuple.C, only the global corrections are actually used and there seems to be no particular reason why o2::tpc::SpaceCharge* sc is a pointer, maybe it is derived from another macro.

But the macro I was actually looking for and which I meant is Detectors/TPC/reconstruction/macro/createTPCSpaceChargeCorrection.C . Here, the AliTPCSpaceCharge3DCalc class is used, so it is not directly affected by the changes, only when we also delete those classes. But it would still be great if you could also change this macro. Also here, the spaceCharge object is a pointer to keep flexibility with grid size and interpolation techniques, but I think in your new implementation this might be redundant and you can make it an object.

There are also two debug functions in createTPCSpaceChargeCorrection.C, one to compare results on the spaceCharge grid and one to compare results on the spline points. In the past, we evaluated the results only for the points for which the corrected point is inside the TPC volume.

matthias-kleiner commented:
> Hi @matthias-kleiner ,
> in the macro generateTPCCorrectionNTuple.C, only the global corrections are actually used and there seems to be no particular reason why o2::tpc::SpaceCharge* sc is a pointer, maybe it is derived from another macro.
> 
> But the macro I was actually looking for and which I meant is Detectors/TPC/reconstruction/macro/createTPCSpaceChargeCorrection.C . Here, the AliTPCSpaceCharge3DCalc class is used, so it is not directly affected by the changes, only when we also delete those classes. But it would still be great if you could also change this macro. Also here, the spaceCharge object is a pointer to keep flexibility with grid size and interpolation techniques, but I think in your new implementation this might be redundant and you can make it an object.
> 
> There are also two debug functions in createTPCSpaceChargeCorrection.C, one to compare results on the spaceCharge grid and one to compare results on the spline points. In the past, we evaluated the results only for the points for which the corrected point is inside the TPC volume.

Hi @ehellbar 
thanks for clarification. 
I will start looking into the createTPCSpaceChargeCorrection.C macro and delete the old SpaceCharge folder now

matthias-kleiner commented:
@ehellbar 
Is there anything else I should modify before making the pull request?
And do you know why "Check PR with clang-format / build (pull_request_target)" always fails?

ehellbar commented:
Hi @matthias-kleiner ,
it all looks fine to me. 
There are a couple of places where the clang-format failed. You can see the diffs in the "Run clang format" tab after you click on the details.

matthias-kleiner commented:
> Hi @matthias-kleiner ,
> it all looks fine to me.
> There are a couple of places where the clang-format failed. You can see the diffs in the "Run clang format" tab after you click on the details.

Hi @ehellbar ,
thanks I will look into it and try to fix it. 
I should probably also squash all my commits into one commit?

ehellbar commented:
Hello @matthias-kleiner ,
yes I think it would be good to squash the commit. The conflict with Detectors/TPC/simulation/src/SpaceCharge.cxx should disappear when you apply your commits on a fresh pull.

wiechula commented on Detectors/TPC/simulation/src/Digitizer.cxx:
```diff
@@ -164,20 +164,33 @@ void Digitizer::flush(std::vector<o2::tpc::Digit>& digits,
 }
 
-void Digitizer::setUseSCDistortions(SpaceCharge* spaceCharge)
+void Digitizer::setUseSCDistortions(SpaceCharge* SpaceCharge)
```
Why is `SpaceCharge` as variable capital?

matthias-kleiner replied:
thanks for noticing it. It should be of course just spaceCharge. I probably changed it by accident

wiechula commented on Detectors/TPC/spacecharge/include/TPCSpacecharge/DataContainer3D.h:
```diff
@@ -0,0 +1,205 @@
+  /// \param iy index in y dimension
+  /// \param iz index in z dimension
+  /// \return returns the index to the data
+  size_t getDataIndex(const size_t ix, const size_t iy, const size_t iz) const
```
can this function be constexpr?

matthias-kleiner replied:
yes I changed it

wiechula reviewed (commented):
Looks fine. Only two very minor comments.

matthias-kleiner commented:
Hello @ehellbar @wiechula
Do you know how I can fix the error in the log:
> fatal error: 'omp.h' file not found

I thought it would be sufficient to put `#ifdef WITH_OPENMP
#include <omp.h>
#endif`everywhere where omp is included?

davidrohr commented:
@matthias-kleiner : Without checking your code, in order to use OpenMP you have to pull in the dependency in the CMake as private as in https://github.com/AliceO2Group/AliceO2/blob/dev/GPU/GPUTracking/CMakeLists.txt#L423:
```
if(OpenMP_CXX_FOUND)
  message(STATUS "GPU: Using OpenMP: ${OpenMP_CXX_SPEC_DATE}")
  # Must be private, depending libraries might be compiled by compiler not understanding -fopenmp
  target_compile_definitions(${targetName} PRIVATE WITH_OPENMP)
  target_link_libraries(${targetName} PRIVATE OpenMP::OpenMP_CXX)
endif()
```

matthias-kleiner commented:
> @matthias-kleiner : Without checking your code, in order to use OpenMP you have to pull in the dependency in the CMake as private as in https://github.com/AliceO2Group/AliceO2/blob/dev/GPU/GPUTracking/CMakeLists.txt#L423:
> 
> ```
> if(OpenMP_CXX_FOUND)
>   message(STATUS "GPU: Using OpenMP: ${OpenMP_CXX_SPEC_DATE}")
>   # Must be private, depending libraries might be compiled by compiler not understanding -fopenmp
>   target_compile_definitions(${targetName} PRIVATE WITH_OPENMP)
>   target_link_libraries(${targetName} PRIVATE OpenMP::OpenMP_CXX)
> endif()
> ```

Hi @davidrohr 
thanks for the comment, but I am already doing it in the CMake as you mentioned.
Could it be that it is something related to ROOT ?:

> input_line_11:91:10: fatal error: 'omp.h' file not found
#include <omp.h>
         ^~~~~~~
Error: /mnt/mesos/sandbox/sandbox/o2-fullci/sw/slc8_x86-64/ROOT/v6-20-08-3/bin/rootcling: compilation failure (/mnt/mesos/sandbox/sandbox/o2-fullci/sw/BUILD/f71ddec13c659766d4c5e61369dd439d397483f8/O2/Detectors/TPC/spacecharge/G__O2TPCSpaceChargecdfa8bf6dd_dictUmbrella.h)
Error copying file (if different) from "/mnt/mesos/sandbox/sandbox/o2-fullci/sw/BUILD/f71ddec13c659766d4c5e61369dd439d397483f8/O2/Detectors/TPC/spacecharge/G__O2TPCSpaceCharge_rdict.pcm" to "/mnt/mesos/sandbox/sandbox/o2-fullci/sw/BUILD/f71ddec13c659766d4c5e61369dd439d397483f8/O2/stage/lib64/G__O2TPCSpaceCharge_rdict.pcm".
gmake[2]: *** [Detectors/TPC/spacecharge/CMakeFiles/O2lib-TPCSpaceCharge.dir/build.make:91: Detectors/TPC/spacecharge/G__O2TPCSpaceCharge.cxx] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:30858: Detectors/TPC/spacecharge/CMakeFiles/O2lib-TPCSpaceCharge.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....

davidrohr commented:
Hm, this might be, I am using omp.h only in cxx not in header files, so it is not seen by the dictionary generation.

shahor02 commented:
Hi, yes, I have also seen this: the omp should not be exposed to rootcling

matthias-kleiner commented:
Hi thanks for your help, I could fix the problem with omp, but now I am getting some `is not a constant expression` errors which I dont understand. 
The first errors seem to come from MathConstants.h which I include:

> /mnt/mesos/sandbox/sandbox/o2-fullci/sw/SOURCES/O2/4618/0/Common/Constants/include/CommonConstants/MathConstants.h:25:31: error: '(1.0e+0f - 1.17549001e-38f)' is not a constant expression
 constexpr float Almost1 = 1.f - Almost0;
                           ~~~~^~~~~~~~~
/mnt/mesos/sandbox/sandbox/o2-fullci/sw/SOURCES/O2/4618/0/Common/Constants/include/CommonConstants/MathConstants.h:26:31: error: '(1.0e+0f / 1.17549001e-38f)' is not a constant expression
 constexpr float VeryBig = 1.f / Almost0;
                           ~~~~^~~~~~~~~
/mnt/mesos/sandbox/sandbox/o2-fullci/sw/SOURCES/O2/4618/0/Common/Constants/include/CommonConstants/MathConstants.h:31:33: error: '(1.8e+2f / 3.14159274e+0f)' is not a constant expression
 constexpr float Rad2Deg = 180.f / PI;
                           ~~~~~~^~~~
/mnt/mesos/sandbox/sandbox/o2-fullci/sw/SOURCES/O2/4618/0/Common/Constants/include/CommonConstants/MathConstants.h:32:30: error: '(3.14159274e+0f / 1.8e+2f)' is not a constant expression
 constexpr float Deg2Rad = PI / 180.f;

The same occurs in Defs.h

> /mnt/mesos/sandbox/sandbox/o2-fullci/sw/SOURCES/O2/4618/0/DataFormats/Detectors/TPC/include/DataFormatsTPC/Defs.h:46:38: error: '(6.2831853071795862e+0 / 1.8e+1)' is not a constant expression
 constexpr double SECPHIWIDTH = TWOPI / 18.;

and later when I define:

> /mnt/mesos/sandbox/sandbox/o2-fullci/sw/SOURCES/O2/4618/0/Detectors/TPC/spacecharge/include/TPCSpaceCharge/PoissonSolver.h:97:40: error: '(1.0e+0 / 6.2831854820251465e+0)' is not a constant expression
   static constexpr DataT INVTWOPI = 1. / o2::constants::math::TwoPI; ///< inverse of 2*pi

davidrohr commented:
These errors are extremely strange, in particular since the MathConstants.h file is compiled in many places, and also it works on macOS.
Perhaps you can try completely without openmp?
Or it would be interesting to test with a different compiler, perhaps there is a problem with gcc 7.3.
I think to investigate, you could try to create a minimum test case that reproduces the issue, e.g. try with a more or less empty CXX file that includes MathConstants.h in the same setup as you currenly have, and let's see if the problem with Almost1 appears there.

matthias-kleiner commented:
> These errors are extremely strange, in particular since the MathConstants.h file is compiled in many places, and also it works on macOS.
> Perhaps you can try completely without openmp?
> Or it would be interesting to test with a different compiler, perhaps there is a problem with gcc 7.3.
> I think to investigate, you could try to create a minimum test case that reproduces the issue, e.g. try with a more or less empty CXX file that includes MathConstants.h in the same setup as you currenly have, and let's see if the problem with Almost1 appears there.

Thanks for your help. I added a dummy class for this purpose, but the errors "is not a constant expression" still persists. It also gives me an error when I just define `static constexpr float DUMMY3 = 1. / 3.1415927;` as a member in the dummy class. Is there anything else I could test to find out where the errors come from?

> Or it would be interesting to test with a different compiler, perhaps there is a problem with gcc 7.3.

How could I test it with a different compiler? Can I do it somehow locally?

davidrohr commented:
To test it locally, you can run it in a docker container. The container is `alisw/slc7-builder` for the O2/o2 CI and `davidrohr/slc8-gpu-builder` for the full CI. Inside, the container, please set the env variable `ALIBUILD_O2_TESTS=1` to have the same setup as in the CI. Then just install and run aliBuild.

davidrohr commented:
For reference, I have tried locally and it happens with GCC versions 7, 9, and 10, so it is not a problem with the GCC in our CI. Having a closer look.

davidrohr commented:
@matthias-kleiner : OK, I managed to understand what happens. Your code is compiled with the compile flag `--frounding-math` which indicates that the rounding behavior may change at runtime because the application may adjust processor registers affecting floating point rounding. That in turn makes floating point division in general non-constexpr, therefore the `TWOPI / 18` is no constexpr.... That GCC error message really should be clearer... Because as is this seems totally like a compiler bug.

Anyhow, the question is where this bogus flag comes from, and it turns out the --frounding-match comes in from cgal. You link to CGAL::CGAL and that automatically sets that compile flag for its interface library, so CMake adds it to everything that is compiled and linked to CGAL. Apparently cgal requires that, or at least otherwise some cgal tests would fail. There are some discussions by other annoyed people as well about this... Personally I consider that quite bogus messing with compile flags automatically... but perhaps we should not hack it away in CGAL but find a clean solution.

The best I can suggest is:
Split your library in 2:
- a base library that links to CGAL
- and the normal library that has your SCD code, and links to the base library

The base library should not include any O2 headers, and the other library should not include any CGAL headers.
In the base library, you have to link to CGAL as
```
PRIVATE_LINK_LIBRARIES CGAL::CGAL
```
and not `PUBLIC` otherwise the flags get propagated further.
Is that feasible for you?

As a side-node, if cgal is required, could you move it from your private CMake to /dependencies and mark it as required there?

matthias-kleiner commented:
> @matthias-kleiner : OK, I managed to understand what happens. Your code is compiled with the compile flag `--frounding-math` which indicates that the rounding behavior may change at runtime because the application may adjust processor registers affecting floating point rounding. That in turn makes floating point division in general non-constexpr, therefore the `TWOPI / 18` is no constexpr.... That GCC error message really should be clearer... Because as is this seems totally like a compiler bug.
> 
> Anyhow, the question is where this bogus flag comes from, and it turns out the --frounding-match comes in from cgal. You link to CGAL::CGAL and that automatically sets that compile flag for its interface library, so CMake adds it to everything that is compiled and linked to CGAL. Apparently cgal requires that, or at least otherwise some cgal tests would fail. There are some discussions by other annoyed people as well about this... Personally I consider that quite bogus messing with compile flags automatically... but perhaps we should not hack it away in CGAL but find a clean solution.
> 
> The best I can suggest is:
> Split your library in 2:
> 
>     * a base library that links to CGAL
> 
>     * and the normal library that has your SCD code, and links to the base library
> 
> 
> The base library should not include any O2 headers, and the other library should not include any CGAL headers.
> In the base library, you have to link to CGAL as
> 
> ```
> PRIVATE_LINK_LIBRARIES CGAL::CGAL
> ```
> 
> and not `PUBLIC` otherwise the flags get propagated further.
> Is that feasible for you?
> 
> As a side-node, if cgal is required, could you move it from your private CMake to /dependencies and mark it as required there?

Many thanks for your help! 
I am not surprised it is CGAL, that causes some problems. I had something similar when using valgrind with CGAL... I will then try to split the library into two parts as you suggested.

davidrohr commented:
Please not also that o2 dataflow defaults disables CGAL, so I guess in that case we should disable building the spacecharge classes.

matthias-kleiner committed:
moved CGAL includes to base class

matthias-kleiner commented:
I could fix the issue with CGAL by moving the CGAL headers and the code to a base library in a class `NearestNeighbour` like @davidrohr suggested. Thanks again for your help :)
However in the class I had to define the members of the class which are objects from CGAL as global objects in the source file to avoid having any CGAL includes in the header file. Is there anything I can do to have a more cleaner solution, or is that solution fine? @wiechula

davidrohr commented on Detectors/TPC/spacecharge/CMakeLists.txt:
```diff
@@ -0,0 +1,50 @@
+
+o2_add_library(TPCSpaceChargeBase
+               TARGETVARNAME targetNameBase
+	       SOURCES src/NearestNeighbour.cxx
```
indentation incorrect.

davidrohr commented on Detectors/TPC/spacecharge/src/NearestNeighbour.cxx:
```diff
@@ -0,0 +1,78 @@
+using K_neighbor_search = CGAL::Orthogonal_k_neighbor_search<Traits>;
+using Tree = K_neighbor_search::Tree;
+
+std::unique_ptr<std::vector<Point_3>> mPoints;
```
You can forward declare a class NearestNeighbourInternal in NearestNeighbour.h and add a member std::unique_ptr<NearestNeighbourInternal> to the NearestNeighbour class.
Then, define the actual ...Internal class inside NearestNeightbour.cxx, and move these global objects in there and create the internal object in the constructor.
Only thing you have to make sure is that both constructor and destructor must be in the .cxx and not in the header. For the destructor, you can declare it in the header and add to the .cxx `NearestNeighbour::~NearestNeighbour() = default;`.

matthias-kleiner replied:
Thanks for the nice explanation :) I moved it to an internal class as suggested

davidrohr reviewed (changes requested):
Nice that this works, you can avoid the global variables as explained below.

matthias-kleiner committed:
moved global objects in NearestNeighbour src file to own class

davidrohr commented:
ok, do I understand correctly that now everything is in?
Then I'd do a test with simulation and reconstruction with distortions and compare it to the previous one.
Can I still use the same spacecharge filles that @ehellbar provided end of last year?

matthias-kleiner commented:
> ok, do I understand correctly that now everything is in?
> Then I'd do a test with simulation and reconstruction with distortions and compare it to the previous one.
> Can I still use the same spacecharge filles that @ehellbar provided end of last year?

yes everything should be in and working and you can still use the same old space charge files.

davidrohr commented:
@TimoWilken : I have seen recently quite often that a CI build fails with an empty log, like here the O2/o2 CI:
```
Builing on: alibuild16.cern.ch at 2020-11-26-08:27:51
No logs found. Please check actual builders log in aurora.
See http://alisw.github.io/infrastructure-pr-testing for more instructions.
```
I know we don't want to expose the full logs since they might contain passwords etc., but is there any chance to show some hint where it failed? Because otherwise I have to ping you so often :|

TimoWilken commented:
For O2/o2, it's:
```
2020-11-26@08:27:51:DEBUG:O2Suite:FairRoot:build_O2_o2: ++ '[' -z ']'
2020-11-26@08:27:51:DEBUG:O2Suite:FairRoot:build_O2_o2: ++ source /mnt/mesos/sandbox/sandbox/o2/sw/slc7_x86-64/asiofi/v0.3.1-1/etc/profile.d/init.sh
2020-11-26@08:27:51:DEBUG:O2Suite:FairRoot:build_O2_o2: /mnt/mesos/sandbox/sandbox/o2/sw/slc7_x86-64/FairMQ/v1.4.27-1/etc/profile.d/init.sh: line 6: /mnt/mesos/sandbox/sandbox/o2/sw/slc7_x86-64/asiofi/v0.3.1-1/etc/profile.d/init.sh: No such file or directory
2020-11-26@08:27:51:ERROR:O2Suite:FairRoot:build_O2_o2: Error while executing /mnt/mesos/sandbox/sandbox/o2/sw/SPECS/slc7_x86-64/FairRoot/v18.4.1-37/build.sh on `alibuild16.cern.ch'.
2020-11-26@08:27:51:ERROR:O2Suite:FairRoot:build_O2_o2: Log can be found in /mnt/mesos/sandbox/sandbox/o2/sw/BUILD/FairRoot-latest/log
2020-11-26@08:27:51:ERROR:O2Suite:FairRoot:build_O2_o2: Please upload it to CERNBox/Dropbox if you intend to request support.
2020-11-26@08:27:51:ERROR:O2Suite:FairRoot:build_O2_o2: Build directory is /mnt/mesos/sandbox/sandbox/o2/sw/BUILD/FairRoot-latest/FairRoot.
```
I'll investigate.

I'm not sure why it didn't find that log, it should have.

I'll add a short error message to GitHub so it shows that it was aliBuild that failed (as opposed to a merge or aliDoctor error, which have their own messages) even when there aren't any logs. I'm not sure there's a better way of doing that without trying to parse the logs at the moment, sorry!

davidrohr commented:
@ktf @TimoWilken : Perhaps as one idea: We already require a cern login for accessing the logs. Is it possible to check the username or the e-groups? Perhaps it would be possible to make the full logs available to infrastructure admins via the web interface?

ktf commented:
Because of limitations of the current S3 / SSO infrastructure, the setup is quite convoluted and ad-hoc. We need another bucket for the full logs, a different "local" user and a separate set of credentials. All of which can be done, but it requires a bit of planning / agreement with CERN/IT. Please open a JIRA ticket with the request and I will try to go through it at some point.

matthias-kleiner committed:
adding more grid sizes for calculation of distortions/corrections

davidrohr commented:
@matthias-kleiner : Could you have a look at the error in the MacOS CI? It fails with the macro createResidualDistortionObject.C, so I think this is related.

davidrohr commented:
@matthias-kleiner : Sorry to bug you again, but could you rebase the PR?

matthias-kleiner commented:
> @matthias-kleiner : Sorry to bug you again, but could you rebase the PR?

Hi @davidrohr no problem,
I will do it later today when I fixed the createResidualDistortionObject.C macro

matthias-kleiner committed:
fixed createResidualDistortionObject.C macro for new space charge classes and moved to space charge folder

davidrohr commented:
@ktf @TimoWilken : This fails with the following error:
```
CMake Error in Detectors/TPC/spacecharge/CMakeLists.txt:
  Imported target "CGAL::CGAL" includes non-existent path

    "/jenkins/workspace/DailyBuilds/DailyO2/sw/20156194/1/slc7_x86-64/GMP/v6.0.0-59/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.
```
I just want to verify whether I understand this correctly: O2 links to the CGAL lib, and CGAL wants to pull in GMP headers, which are not present. I assume this is the case, since the CI used a binary package for CGAL and GMP is not provided, since GMP is only a build requirement for CGAL.
So the fix for this should be to make GMP a normal requirement instead of a build_requirement of CGAL. Could you confirm?

davidrohr commented:
@matthias-kleiner : Could you apply clang-format :)

matthias-kleiner committed:
clang format

davidrohr commented:
@matthias-kleiner : The macro is now failing in all CI instances, not only MacOS, weird. Perhaps it is related to the problem we had before that some tests were simply not running.
Anyhow, I think the CI right now tests the macro in the interpreted mode. Can it actually run interpreted, or does it need to be compiled? I think if so you have to indicated that in the CMake when you add the macro test.

matthias-kleiner commented:
> @matthias-kleiner : The macro is now failing in all CI instances, not only MacOS, weird. Perhaps it is related to the problem we had before that some tests were simply not running.
> Anyhow, I think the CI right now tests the macro in the interpreted mode. Can it actually run interpreted, or does it need to be compiled? I think if so you have to indicated that in the CMake when you add the macro test.

The calculateDistortionsCorrections.C macro needs to be compiled. The other createResidualDistortionObject.C macro should also run in interpreted mode. I will then indicate it with "COMPILE_ONLY"

matthias-kleiner committed:
changed macro calculateDistortionsCorrections.C to compile only

davidrohr commented:
@ktf @TimoWilken : Could you check what is the problem with CGAL and GMP. After alisw/alidist#2648 GMP should be a direct build dependency of O2, So how can it be that the GMP include folder does not exist? :
```
CMake Error in Detectors/TPC/spacecharge/CMakeLists.txt:
  Imported target "CGAL::CGAL" includes non-existent path

    "/jenkins/workspace/DailyBuilds/DailyO2/sw/20156194/1/slc7_x86-64/GMP/v6.0.0-59/include"
```
I have tried this locally in the slc7-builder container but cannot reproduce it.

TimoWilken commented:
@davidrohr @ktf That includes something under `/jenkins/workspace`, but the CI builders are working under `/mnt/mesos/sandbox/sandbox`. I'd guess there is something wrong with the Jenkins build?

davidrohr commented:
@matthias-kleiner : Before I was definint the gridsize for the digitizer workflow via `--gridSize=65,180,65`. This option no longer exists. How do I set the grid size now?

matthias-kleiner commented:
> @matthias-kleiner : Before I was definint the gridsize for the digitizer workflow via `--gridSize=65,180,65`. This option no longer exists. How do I set the grid size now?

the gridsize is now set to a default of 129,180,129. There is currently no option to set it.

davidrohr commented:
@matthias-kleiner : I have retried without the gridsize setting, but the digitization output for the TPC looks broken. Below are 2 visualizations of a single pp collision before and after this PR digitized with
```
o2-sim-digitizer-workflow --TPCtriggered --distortionType 2 --initialSpaceChargeDensity=../InputSCDensityHistograms_8000events.root,inputSCDensity3D_8000_0 --incontext collisioncontext.root --tpc-lanes=1 --onlyDet TPC
```
using the SCD files provided by Ernst some time ago.
<img width="1920" alt="before" src="https://user-images.githubusercontent.com/9131638/100884189-79aa6d00-34b1-11eb-8a5a-c2e9f873037e.png">
<img width="1920" alt="after" src="https://user-images.githubusercontent.com/9131638/100884226-8333d500-34b1-11eb-89c2-661256e980a7.png">

matthias-kleiner commented:
> @matthias-kleiner : I have retried without the gridsize setting, but the digitization output for the TPC looks broken. Below are 2 visualizations of a single pp collision before and after this PR digitized with
> 
> ```
> o2-sim-digitizer-workflow --TPCtriggered --distortionType 2 --initialSpaceChargeDensity=../InputSCDensityHistograms_8000events.root,inputSCDensity3D_8000_0 --incontext collisioncontext.root --tpc-lanes=1 --onlyDet TPC
> ```
> 
> using the SCD files provided by Ernst some time ago.
> <img alt="before" width="1920" src="https://user-images.githubusercontent.com/9131638/100884189-79aa6d00-34b1-11eb-8a5a-c2e9f873037e.png">
> <img alt="after" width="1920" src="https://user-images.githubusercontent.com/9131638/100884226-8333d500-34b1-11eb-89c2-661256e980a7.png">

Okay, thanks for doing this check. I will look into this and see if I can figure out what goes wrong

matthias-kleiner committed:
fixed z coordinate in distortElectron

matthias-kleiner commented:
@davidrohr could you please do the same check again? I just found a bug where I set the z coordinate equal to the y coordinate after the distortion. If this doesnt fix the issue, I could look tomorrow more into it.

davidrohr reviewed (dismissed):
@matthias-kleiner : Thx, with this fix it looks OK, and also the reconstruction with the correction looks like before. So if the CI finally passes, we can finally merge.
Could you just quickly comment on the gridSize parameter? I have seen that also in the macro to create the tpcTransform object, the gridSize parameter is gone. So I have just left it out for my tests. Does it not exist anymore, or is the default just ok, or shall I set it somehow else?

matthias-kleiner commented:
> @matthias-kleiner : Thx, with this fix it looks OK, and also the reconstruction with the correction looks like before. So if the CI finally passes, we can finally merge.
> Could you just quickly comment on the gridSize parameter? I have seen that also in the macro to create the tpcTransform object, the gridSize parameter is gone. So I have just left it out for my tests. Does it not exist anymore, or is the default just ok, or shall I set it somehow else?

Since the grid size is templated now in the spacecharge class, the easiest solution for now was to just set it to a default value during the simulations and in the macros. Of course the correct default value of the grid size still has to be determined, but gridR=129, gridPhi180, gridZ=129 should already be a good starting point. The grid size is just hardcoded for example in the digitizer `using SC = SpaceCharge<double, 129, 129, 180>;`. If you want to change the grid size, the only way for now is to change the hardcoded values in Digitizer.h, TPCDigitizerSpec.cxx, and the macros. 

Of course it would be better to be able to set the grid size and pass somehow the template parameters of the grid or choose between different presets of grid sizes, but I am not sure if there is an elegant way to implement this in the Digitizer.

sawenzel commented on Detectors/TPC/spacecharge/include/TPCSpaceCharge/SpaceChargeHelpers.h:
```diff
@@ -0,0 +1,269 @@
+  /// analytical potential
+  std::function<DataT(DataT, DataT, DataT)> mPotentialFunc = [& mParA = mParA, &mParB = mParB, &mParC = mParC](const DataT z, const DataT r, const DataT phi) {
+    const DataT zz = std::abs(z);
+    return -mParA * (std::pow((-r + 254.5 + 83.5), 4) - 338.0 * std::pow((-r + 254.5 + 83.5), 3) + 21250.75 * std::pow((-r + 254.5 + 83.5), 2)) * std::cos(mParB * phi) * std::cos(mParB * phi) * std::exp(-1 * mParC * (zz - 125) * (zz - 125));
```
depending on where or how often this function is used during runtime, I'd like to point out that constructs  like `std::pow((-r + 254.5 + 83.5), 4)` are unfortunately inducing a function call to pow and therefore slow. You can check with https://godbolt.org/ that this is only optimized away with `-ffast-math` which we are not using.

matthias-kleiner replied:
thanks for the comment :) I didnt knew that about `std::pow`. But since these functions are only used for "debugging/testing" anyway, it does not really matter how fast the evaluation is.

sawenzel commented on Detectors/TPC/spacecharge/src/PoissonSolver.cxx:
```diff
@@ -0,0 +1,1497 @@
+  int tnRRow = iOne == 1 ? Nr : Nr / iOne + 1;
+  int tnZColumn = jOne == 1 ? Nz : Nz / jOne + 1;
+
+  std::vector<DataT> coefficient1(Nr);
```
I am not quite getting the rational why the GRID size is a template argument everywhere. But since it is, there is no need to use dynamic memory in many places. `std::array<DataT, Nr> coefficent` would be a more sensible choice.

matthias-kleiner replied:
You are right. The grid size could also be just a member in the different classes and then one could use a std::vector everywhere, which would make the code a lot easier. But at the beginning I thought it would make sense to use arrays instead of vectors and to ensure that one uses the correct grid sizes (there are only specific grid sizes which can be used) and that these classes can only be used with the same grid size. I might remove the templates in the future, but this would require some work...

Indeed it makes sense to use an array there`std::array<DataT, Nr> coefficent`, but since this vector is passed to several member functions and in a different function in the code the same coefficient array is not depending on the templated grid size `std::vector<DataT> coefficient1(tnRRow)` it has to be a vector there. I could probably do some additional templating, to use vectors and arrays where appropriate.

sawenzel reviewed (dismissed):
It's a nice PR. Please consider my comments for a future polishing step.

davidrohr commented:
shit, this breaks O2-dataflow defaults, because they disable CGAL.
What do we do. We could remove the required for CGAL, but that will probably be complicated since then we have do disable compiling this code and all that depends on it dynamically depending on whether CGAL is found.
Perhaps I would make it depend on BUILD_SIMULATION:
- Add CGAL as simulation dependency in O2SimulationDependencies.cmake.
- And then disable the respective libraries when simulation is not present. Might probably be easier since then we don't have to distinguish in the digitizer?

matthias-kleiner commented:
> shit, this breaks O2-dataflow defaults, because they disable CGAL.
> What do we do. We could remove the required for CGAL, but that will probably be complicated since then we have do disable compiling this code and all that depends on it dynamically depending on whether CGAL is found.
> Perhaps I would make it depend on BUILD_SIMULATION:
> 
>     * Add CGAL as simulation dependency in O2SimulationDependencies.cmake.
> 
>     * And then disable the respective libraries when simulation is not present. Might probably be easier since then we don't have to distinguish in the digitizer?

Hi @davidrohr, if the CGAL package causes too much trouble, I could try and see if I can remove it in the space charge class. CGAL is used only in one specific function and I think it might be possible/equally fast to remove CGAL and modify the function where it is used. I can check this and see if its reasonable without CGAL.

davidrohr commented:
> Hi @davidrohr, if the CGAL package causes too much trouble, I could try and see if I can remove it in the space charge class. CGAL is used only in one specific function and I think it might be possible/equally fast to remove CGAL and modify the function where it is used. I can check this and see if its reasonable without CGAL.

OK, can you check that quickly? If it is too much effort, I just discussed with @sawenzel that we could add CGAL as requirement also for o2-dataflow.

davidrohr commented:
But if that would mean copy&pasting much code, we can also just enforce having CGAL. So perhaps have a quick look what needs to be done and then report back. In the meantime, I am checking what would need to be done to make the CGAL optional.

matthias-kleiner commented:
> But if that would mean copy&pasting much code, we can also just enforce having CGAL. So perhaps have a quick look what needs to be done and then report back. In the meantime, I am checking what would need to be done to make the CGAL optional.

I use CGAL actually only for finding the nearest neighbour of some scattered datapoints and then apply some iterative method to approach some query point. I think i could just skip the nearest neighbour searching and start just at some point in the middle of the grid... There would be actually less code than before, but I have to check how fast it is and if the result will be the same. I can test this later today

davidrohr commented:
ok, thx, actually we should have thought about that before we implemented the other workarounds...

matthias-kleiner commented:
> ok, thx, actually we should have thought about that before we implemented the other workarounds...

Hi @davidrohr, it looks like I can get rid of the CGAL package and use some slightly different approach in the function where CGAL is used. 
The resulting distortions are basically the same and the performance is actually a little bit better.
I am really sorry for all the trouble and work you had with this. 

I will then remove all the CGAL stuff from my classes and push it as soon as possible.

davidrohr commented:
@matthias-kleiner : Ok, great.
And no worries, I guess we all learned something in this endeavor :)

matthias-kleiner committed:
removed CGAL package and replaced nearest neighbour searching by approximation of nearest neighbour
